codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/HttpBindingProtocolGenerator.java

```diff
@@ -114,16 +169,136 @@ private void generateOperationSerializer(
             GenerationContext context,
             OperationShape operation
     ) {
+        generateOperationSerializerMiddleware(context, operation);
         generateOperationHttpBindingSerializer(context, operation);
+        generateOperationDocumentSerializer(context, operation);
+        addOperationDocumentShapeBinders(context, operation);
+    }
+
+    protected abstract void generateOperationDocumentSerializer(GenerationContext context, OperationShape operation);
+
+    /**
+     * Adds the top-level shapes from the operation that bind to the body document that require serializer functions.
+     *
+     * @param context   the generator context
+     * @param operation the operation to add document binders from
+     */
+    private void addOperationDocumentShapeBinders(GenerationContext context, OperationShape operation) {
+        Model model = context.getModel();
+
+        /*
+         * Walk and members shapes to the list that will require serializer functions
+         */
+        List<HttpBinding> bindings = new ArrayList<>(model.getKnowledge(HttpBindingIndex.class)
+                .getRequestBindings(operation).values());
```

Comment:
Why convert this to a list? All you're doing with it is iterating over it, which you can do just as well with a set.